src/pages/Admin/VenueOwnersModal.tsx

```diff
@@ -1,66 +1,101 @@
-import React, { useMemo, useState, useCallback } from "react";
-import { Modal } from "react-bootstrap";
-import { useFirestoreConnect } from "react-redux-firebase";
-import { useSelector } from "hooks/useSelector";
-import Fuse from "fuse.js";
+import React, { useMemo, useState, useCallback, useEffect } from "react";
+import { useFirestore } from "react-redux-firebase";
+import { Modal, FormControl } from "react-bootstrap";
 import { debounce } from "lodash";
-import { FormControl } from "react-bootstrap";
-import { User } from "types/User";
-import "./VenueOwnerModal.scss";
+
+import { DEFAULT_PARTY_NAME, DEFAULT_PROFILE_IMAGE } from "settings";
 import { AnyVenue } from "types/Firestore";
+import { User } from "types/User";
 import { WithId } from "utils/id";
-import { DEFAULT_PARTY_NAME, DEFAULT_PROFILE_IMAGE } from "settings";
 import { addVenueOwner, removeVenueOwner } from "api/admin";
 
+import "./VenueOwnerModal.scss";
+
+interface PartitionedOwnersOthers {
+  owners: WithId<User>[];
+  others: WithId<User>[];
+}
+const makePartitionOwnersFromOthersReducer = (ownerIds: string[]) => (
+  { owners, others }: PartitionedOwnersOthers,
+  user: WithId<User>
+) => {
+  if (ownerIds.includes(user.id)) {
+    return { owners: [...owners, user], others };
+  } else {
+    return { owners, others: [...others, user] };
+  }
+};
+
+const emptyPartition: PartitionedOwnersOthers = {
+  owners: [],
+  others: [],
+};
+
+const makePartyNameFilter = (searchText: string) => (user: WithId<User>) =>
+  user.partyName?.toLowerCase()?.includes(searchText.toLowerCase());
+
 interface VenueOwnersModalProps {
   visible: boolean;
   venue: WithId<AnyVenue>;
   onHide?: () => void;
 }
 
-export const VenueOwnersModal: React.FC<VenueOwnersModalProps> = (props) => {
-  const { visible, onHide, venue } = props;
+export const VenueOwnersModal: React.FC<VenueOwnersModalProps> = ({
+  visible,
+  onHide,
+  venue,
+}) => {
+  // Fetch all users the first time this component loads
+  // @debt reading every user is obviously bad.
+  const firestore = useFirestore();
+  useEffect(() => {
+    firestore
+      .collection("users")
+      .get()
+      .then((result) =>
+        result.docs.map<WithId<User>>(
+          (doc) => ({ ...doc.data(), id: doc.id } as WithId<User>) // TODO: be less hacky with types here?
+        )
+      )
+      .then(setAllUsers);
+  }, [firestore]);
 
-  // @debt reading every user into redux is obviously bad.
-  // A better solution is to create an index on partyName and a firebase function to search
-  useFirestoreConnect({
-    collection: "users",
-    storeAs: "allUsers",
-  });
```

Comment:
Calling this out, as this is my biggest 'hrmmm...' of my changes.

The `@debt` comment already existed, yet the previous code was using `useFirestoreConnect`, which I believe was causing it to be run _many_ times, as higher level components re-rendered, etc.

It could probably be done better, but this seemed the simplest 'isolated change' way to go about it for now. May make sense to clean this up as part of a larger data fetching architecture re-think in the future.

Clicking through, even though it is only fetched the first time the component is rendered, the app seems to continue to function as expected. That said.. if in future we fix higher level components to reduce their re-renders, I'm not 100% certain this will still be the case.

Open to thoughts/alternatives.

Reply:
Agreed, it's a great find. More broadly it speaks to a problem with the "go get all the data in this collection from a distant datasource, and I'll filter it client-side" approach. Firebase really falls down here, and I predict it'll trigger a complete replatforming at some point in the future.

We should approach Firebase account execs & work with them on solutions because I have struggled so far on engineering solutions to this problem. I want to stay on Firebase as long as possible - it's amazing to not be woken up by infra pages.

Reply:
Makes sense. There are probably some aspects of Firebase that i'm not particularly deeply aware of yet RE: performance tuning, best practices/patterns, etc too, so that could extend our ability to keep using it.

Reply:
There may also be some underlying bugs in the libraries we are using and/or in how we are using them, eg.

- https://github.com/sparkletown/sparkle/pull/716#issuecomment-727189462